### DISCUSSION TITLE
Why does set_text not work?

### DISCUSSION THREAD

**Discussion by shakamoushie** [zauberzeug/nicegui#268] [Q&A]
In the code below, why is it not possible to set the value of the `wch_calc` variable to the `e.value` of the `ui.tree` widget with `on_select=lambda e: wch_calc.set_text(e.value)`?

Because `wch_calc` cannot be set, it is not being processed by the `if wch_calc == ''` statement.

Any help is appreciated...

Thanks in advance.

```py
from nicegui import ui

wch_calc = ''

def DisplayCalcOverview():
    global wch_calc
    return f"<h1>You chose {wch_calc.upper()}</h1>"

@ui.page('/main_page')
def main_page():
    global wch_calc
    with ui.left_drawer(top_corner=False, bottom_corner=False).style('background-color: #F8EDE3').props('width=260 bordered'):
        with ui.row():
            with ui.column():
                with ui.row():
                    ui.html("<p style='font-size: 24px;'>Calculators:</p>")

                ui.tree([
                            {'id': 'Drug Dose', 'children': [{'id': 'BSA'}]},
                            {'id': 'Oncology', 'children': [{'id': 'Recist S'}, {'id': 'Recist M'}, {'id': 'Recist A'}]},
                            {'id': 'Gastroenterology', 'children': [{'id': 'Mayo Score'}, {'id': 'Harvey Bradshaw'}, {'id': 'SCCAI'}, {'id': 'SESCD'}]},
                        ], label_key='id', on_select=lambda e: wch_calc.set_text(e.value))
                ui.notify(wch_calc)

        if wch_calc == '':
            ui.label('No Calc chosen')
        else:
            htmlstr = DisplayCalcOverview(wch_calc)
            ui.html(htmlstr)

ui.button("Click for Main Page", on_click=lambda: ui.open('main_page'))

ui.run()
```

**Comment by falkoschindler** [comment]
`wch_calc` is a Python string. Therefore it has no `set_text` method. Unfortunately there seems to be no (sane) way to alter a plain variable within a lambda expression...
But you can define a global `set_wch_calc` function and register it like this: `on_select=lambda e: set_wch_calc(e.value)`.

**Comment by shakamoushie** [comment]
Thanks @falkoschindler for the prompt response. I amended the code as further below.

I have a follow-up question. Even though I am able to assign the value (e.value of the ui.tree) to the variable (wch_calc) in the function set_wch_calc, its value is not retained in wch_calc even after ui.update. The statement (after the ui.tree statement) ui.label(f"wch_calc: {wch_calc}") has no value.
```py
from nicegui import ui

wch_calc = ''
def set_wch_calc(vval):
    global wch_calc

    wch_calc = vval
    ui.update

def DisplayCalcOverview():
    global wch_calc
    return f"<h1>You chose {wch_calc.upper()}</h1>"

@ui.page('/main_page')
def main_page():
    global wch_calc

    with ui.left_drawer(top_corner=False, bottom_corner=False).style('background-color: #F8EDE3').props('width=260 bordered'):
        with ui.row():
            with ui.column():
                with ui.row():
                    ui.html("<p style='font-size: 24px;'>Calculators:</p>")

                ui.tree([
                            {'id': 'Drug Dose', 'children': [{'id': 'BSA'}]},
                            {'id': 'Oncology', 'children': [{'id': 'Recist S'}, {'id': 'Recist M'}, {'id': 'Recist A'}]},
                            {'id': 'Gastroenterology', 'children': [{'id': 'Mayo Score'}, {'id': 'Harvey Bradshaw'}, {'id': 'SCCAI'}, {'id': 'SESCD'}]},
                        ], label_key='id', on_select=lambda e: set_wch_calc(e.value))
                
        ui.label(f"wch_calc: {wch_calc}")
        if wch_calc == '':
            ui.label('No Calc chosen')
        else:
            htmlstr = DisplayCalcOverview(wch_calc)
            ui.html(htmlstr)

ui.button("Click for Main Page", on_click=lambda: ui.open('main_page'))
ui.run()
```

**Reply by falkoschindler** [reply]
First of all: `ui.update` is not called (there are brackets missing). But it wouldn't do anything anyway. Changing a global variable won't magically update the UI. I know, this is how streamlit works. But this breaks the normal flow of code execution and leads to a bunch of other [complications](https://github.com/zauberzeug/nicegui/issues/1).

In NiceGUI you can use [bindings](https://nicegui.io/reference#bindings) to automatically update UI elements whenever some model value changes. Or - what we also quite often do - define your own `update` function that sets the label text and other properties of all relevant UI elements whenever it is called. Then you can call it in `set_wch_calc` or wherever you handle state changes.

**Reply by shakamoushie** [reply]
Sorry @falkoschindler, for trying your patience. I read the documentation and still couldn't figure it out. My understanding is as follows:

If I have a toggle as defined below:
```py
def Asgn(e):
    global myvar
    myvar = e

myvar = ui.toggle([1, 2, 3], on_change=lambda e: Asgn(e.value)).value
```

1. Within `Asgn`, I can do both, (a) access and use the value of `myvar` and (b) pass `myvar` into any other ui widget
2. If I add the following statement
    ```py
    ui.timer(1, lambda: ui.notify(f"toggle_value:=> {myvar}"))
    ```
    just below the `ui.toggle` statement in the main section of my code, I can see the toggle value notified onscreen
3. However, instead of the `ui.timer` statement, if I wanted to add these following statements:
   ```py
    if myvar == 2:
       ui.notify("You chose number 2")
   ```
    below the `ui.toggle` statement, it would not work

   And this is precisely what I want to do. If in point 3, I wanted to say `if myvar == 2` then update its value into the database, I cant

   Yes, I have used streamlit extensively, but they have limitations such as script rerun and lack of cosmetic options for widgets, etc... 

   I merely wish to know how could I access the value of `myvar` (not through a ui element, but just on its own, like in point 3), if that statement were to be placed below the `ui.toggle` statement (and not within the `Asgn` callback)
  
   It will greatly help if you could share demo code to do this whenever possible. Apologies once again for not being too clear.

Cheers

**Reply by falkoschindler** [reply]
Your first scenario is as follows:

```py
def Asgn(e):
    global myvar
    myvar = e

myvar = ui.toggle([1, 2, 3], on_change=lambda e: Asgn(e.value)).value
ui.timer(1, lambda: ui.notify(f"toggle_value:=> {myvar}"))
```

The `ui.timer` statement creates a timer that executes the callback `lambda: ui.notify(f"toggle_value:=> {myvar}")` every second. So the value of the variable `myvar` is shown repeatedly. So far so good.

Your second scenario is as follows:

```py
def Asgn(e):
    global myvar
    myvar = e

myvar = ui.toggle([1, 2, 3], on_change=lambda e: Asgn(e.value)).value
if myvar == 2:
    ui.notify("You chose number 2")
```

Now the if-statement an the notification is _not_ a callback, but is executed once - even before the UI is started with `ui.run`. At this point the variable `myvar` is still `None` and therefore you don't get any output. Later this line is never evaluated again, so the value of `myvar` doesn't matter anymore.

**Reply by shakamoushie** [reply]
Yes, I got that @falkoschindler, thank you. 

I am specifically only referring to the 2nd scenario below. 

So, if I understand correctly, if I need to update a database with the value of myvar, it will need to be only from within the Asgn callback, right?

Now, consider this: What if I had multiple widgets on the screen, like one has in any database-related data entry form? This means I would have the widget values only accessible from within each widget's callback function, right? This means I would have to update each widget value into the database independently only when that widget has a callback? Is this the correct understanding?

**Reply by falkoschindler** [reply]
> What if I had multiple widgets on the screen, like one has in any database-related data entry form? This means I would have the widget values only accessible from within each widget's callback function, right? This means I would have to update each widget value into the database independently only when that widget has a callback?

Sorry, I don't understand what you are saying or asking.

**Reply by shakamoushie** [reply]
Ok, I will rephrase that @falkoschindler,

Referring to the the 2nd scenario above, let's say I have 3 ui.toggle widgets, with variables (myvar1, myvar2 & myvar3) on screen. Each widget will have its own callback (Asgn1, Asgn2 & Asgn3) respectively. All these 3 widget values need to put into a database as part of a single record.

How do I update the database if the values of myvar1, myvar2 and myvar3 are only accessible from within their own callback and not outside it? 

If the user enters a value in the 1st ui.toggle (myvar1), program control will go to the Asgn1 callback after entry, where I can access the value of myvar1. I will have to have database update code within Asgn1 to update the database with only the value of myvar1 (as myvar2 & myvar3) are not accessible from within Asgn1. 

And then the same thing is applicable when the user enters the value for myvar2 and myvar3?

In programming languages I have gone through, there is a way to control the scope visibility of a variable, so that its values can only be accessed from certain parts of a program. In nicegui, do you have to have a class for accessing multiple variable values at one place? 

Do you have a simple program that I can refer to that has 3-4 different types of widgets, that need to go into a database as part of a single record?  

Thanks in advance.

**Reply by falkoschindler** [reply]
> How do I update the database if the values of myvar1, myvar2 and myvar3 are only accessible from within their own callback and not outside it?

Since `myvar1`, `myvar2` and `myvar3` are global, you can access them from any function:

```py
x = 1.0
y = 2.0

def set_x(value):
    global x, y
    x = value
    ui.notify(f'Writing x = {x}, y = {y} to database')

def set_y(value):
    global x, y
    y = value
    ui.notify(f'Writing x = {x}, y = {y} to database')

ui.number('x', value=x, on_change=lambda e: set_x(e.value))
ui.number('y', value=y, on_change=lambda e: set_y(e.value))
```

Usually you try to avoid using the global scope too much. It is good practice to move model state into a dedicated class. Besides holding the state it can also contain some convenience methods like `set_x` and `set_y`:

```py
class Model:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def set_x(self, value):
        self.x = value
        ui.notify(f'Writing x = {self.x}, y = {self.y} to database')

    def set_y(self, value):
        self.y = value
        ui.notify(f'Writing x = {self.x}, y = {self.y} to database')

model = Model(1.0, 2.0)

ui.number('x', value=model.x, on_change=lambda e: model.set_x(e.value))
ui.number('y', value=model.y, on_change=lambda e: model.set_y(e.value))
```

But NiceGUI comes with a simpler method for updating the model: We can _bind_ UI elements to model attributes, so we don't need those setters. Now the `on_change` event is only used for writing the model to the database:

```py
class Model:
    def __init__(self, x, y):
        self.x = x
        self.y = y

model = Model(1.0, 2.0)

def write_to_database():
    ui.notify(f'Writing x = {model.x}, y = {model.y} to database')

ui.number('x', value=model.x, on_change=write_to_database).bind_value_to(model, 'x')
ui.number('y', value=model.y, on_change=write_to_database).bind_value_to(model, 'y')
```

And once we are here, we can make use of Python's dataclasses that are designed as pure data containers and shorten the code once more:

```py
from dataclasses import dataclass

@dataclass
class Model:
    x: float
    y: float

model = Model(1.0, 2.0)

def write_to_database():
    ui.notify(f'Writing x = {model.x}, y = {model.y} to database')

ui.number('x', value=model.x, on_change=write_to_database).bind_value_to(model, 'x')
ui.number('y', value=model.y, on_change=write_to_database).bind_value_to(model, 'y')
```

**Reply by shakamoushie** [reply]
Hi @falkoschindler, this is precisely what I wanted to know. Thank you so much. You are the best. Sorry for troubling you so much. 

God bless :-)
